build-docker-image/action.yml

```diff
@@ -0,0 +1,83 @@
+name: Build docker image
+description: Build a docker image and push it to Github container registry
+
+inputs:
+  docker_repository:
+    description: Repository name in the container registry. e.g. ghcr.io/dfe-digital/register-trainee-teachers
+    required: true
+  github_token:
+    description: Default Github token retrieved via secrets.github_token
+    required: true
+  snyk_token:
+    description: Snyk token for image vulnerability scanning
+  reuse_cache:
+    description: Use previously built docker image layers to improve build time. Set to false to refresh image.
+    default: true
+    type: boolean
+
+runs:
+  using: composite
+  steps:
+    - name: Login to GitHub Container Registry
+      uses: docker/login-action@v2
+      with:
+        registry: ghcr.io
+        username: ${{ github.repository_owner }}
+        password: ${{ inputs.github_token }}
+
+    - name: Set environment variables (push)
+      if: github.event_name == 'push'
+      shell: bash
+      run: |
+        GIT_BRANCH=${GITHUB_REF##*/}
+        echo "BRANCH_TAG=$GIT_BRANCH" >> $GITHUB_ENV # GIT_BRANCH will be main for refs/heads/main
+        echo "IMAGE_TAG=${{ github.sha }}" >> $GITHUB_ENV
+
+    - name: Set environment variables (pull_request)
+      if: github.event_name == 'pull_request'
+      shell: bash
+      run: |
+        GIT_BRANCH=${GITHUB_HEAD_REF##*/}
+        echo "BRANCH_TAG=$GIT_BRANCH" >> $GITHUB_ENV
+        echo "IMAGE_TAG=${{ github.event.pull_request.head.sha }}" >> $GITHUB_ENV
+
+    - name: Build docker image using cache
+      uses: docker/build-push-action@v4
+      if: inputs.reuse_cache
+      with:
+        tags: |
+          ${{ inputs.docker_repository }}:${{ env.IMAGE_TAG }}
+          ${{ inputs.docker_repository }}:${{ env.BRANCH_TAG }}
+        push: false
+        load: true
+        cache-to: type=inline
+        cache-from: |
+          type=registry,ref=${{ inputs.docker_repository }}:main
+          type=registry,ref=${{ inputs.docker_repository }}:${{ env.IMAGE_TAG }}
+          type=registry,ref=${{ inputs.docker_repository }}:${{ env.BRANCH_TAG }}
+        build-args: COMMIT_SHA=${{ env.IMAGE_TAG }}
+
+    - name: Build docker image without reusing cache
+      uses: docker/build-push-action@v4
+      if: ${{! inputs.reuse_cache}}
+      with:
+        tags: |
+          ${{ inputs.docker_repository }}:${{ env.IMAGE_TAG }}
+          ${{ inputs.docker_repository }}:${{ env.BRANCH_TAG }}
+        push: false
+        load: true
+        cache-to: type=inline
+        build-args: COMMIT_SHA=${{ env.IMAGE_TAG }}
+
+    - name: Run Snyk to check Docker image for vulnerabilities
+      uses: snyk/actions/docker@master
+      if: inputs.snyk_token != ''
+      env:
+        snyk_token: ${{ inputs.snyk_token }}
+      with:
+        image: ${{ inputs.docker_repository }}:${{ env.IMAGE_TAG }}
+        args: --file=Dockerfile --severity-threshold=high --exclude-app-vulns
```

Comment:
Worth having an input to override the working directory?

Reply:
Do you have a use case for it or could we add it when necessary?

Reply:
added input to override default